File: CursorEditor.java
// CursorEditor.java
import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.*;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class CursorEditor {
    @Parameter(names = "--new")
    private boolean newFlag;
    @Parameter(names = "--size")
    private int size = 32;
    @Parameter(names = "--load")
    private String loadFile;
    @Parameter(names = "--save")
    private String saveFile;
    @Parameter(names = "--export")
    private String exportFile;
    @Parameter(names = "--hotspot")
    private String hotspot;
    @Parameter(names = "--color")
    private String color;
    @Parameter(names = "--pixel")
    private String pixel;
    @Parameter(names = "--fill")
    private String fill;
    @Parameter(names = "--preview")
    private boolean preview;

    static class EditorData {
        int size;
        int[][] pixels;
        int[] hotspot;
        int[] palette;
    }

    private int[][] pixels;
    private int currentColor = 1;
    private int[] hotspotPos = {0, 0};
    private int[] palette = {1, 2, 3, 4, 5, 6, 7, 8};
    private Map<Integer, int[]> colorMap = new HashMap<>();

    public CursorEditor() {
        colorMap.put(0, new int[]{255, 255, 255, 0});
        colorMap.put(1, new int[]{0, 0, 0, 255});
        colorMap.put(2, new int[]{255, 255, 255, 255});
        colorMap.put(3, new int[]{255, 0, 0, 255});
        colorMap.put(4, new int[]{0, 255, 0, 255});
        colorMap.put(5, new int[]{0, 0, 255, 255});
        colorMap.put(6, new int[]{255, 255, 0, 255});
        colorMap.put(7, new int[]{255, 0, 255, 255});
        colorMap.put(8, new int[]{0, 255, 255, 255});
    }

    private void initPixels() {
        pixels = new int[size][size];
        for (int i = 0; i < size; i++) {
            Arrays.fill(pixels[i], 0);
        }
    }

    private void setPixel(int x, int y, int colorIdx) {
        if (x >= 0 && x < size && y >= 0 && y < size) {
            pixels[y][x] = colorIdx;
        }
    }

    private int getPixel(int x, int y) {
        if (x >= 0 && x < size && y >= 0 && y < size) {
            return pixels[y][x];
        }
        return 0;
    }

    private void fill(int x, int y, int colorIdx) {
        int target = getPixel(x, y);
        if (target == colorIdx) return;
        Stack<int[]> stack = new Stack<>();
        stack.push(new int[]{x, y});
        Set<String> visited = new HashSet<>();
        while (!stack.isEmpty()) {
            int[] pos = stack.pop();
            int cx = pos[0], cy = pos[1];
            String key = cx + "," + cy;
            if (visited.contains(key)) continue;
            visited.add(key);
            if (getPixel(cx, cy) == target) {
                setPixel(cx, cy, colorIdx);
                int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};
                for (int[] d : dirs) {
                    int nx = cx + d[0], ny = cy + d[1];
                    if (nx >= 0 && nx < size && ny >= 0 && ny < size) {
                        stack.push(new int[]{nx, ny});
                    }
                }
            }
        }
    }

    private BufferedImage toImage() {
        BufferedImage img = new BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB);
        for (int y = 0; y < size; y++) {
            for (int x = 0; x < size; x++) {
                int[] c = colorMap.get(pixels[y][x]);
                if (c == null) c = new int[]{0, 0, 0, 0};
                img.setRGB(x, y, new Color(c[0], c[1], c[2], c[3]).getRGB());
            }
        }
        return img;
    }

    private void exportCur(String filename) throws IOException {
        BufferedImage img = toImage();
        String pngName = filename.replace(".cur", ".png");
        ImageIO.write(img, "png", new File(pngName));
        System.out.println("\u001B[32mЭкспортировано в " + filename + " (как PNG)\u001B[0m");
    }

    private EditorData toData() {
        EditorData data = new EditorData();
        data.size = size;
        data.pixels = pixels;
        data.hotspot = hotspotPos;
        data.palette = palette;
        return data;
    }

    private void fromData(EditorData data) {
        this.size = data.size;
        this.pixels = data.pixels;
        this.hotspotPos = data.hotspot;
        this.palette = data.palette;
    }

    private void preview() {
        System.out.println("\u001B[36mПредпросмотр курсора (" + size + "x" + size + "), hotspot: " + hotspotPos[0] + "," + hotspotPos[1] + "\u001B[0m");
        String[] chars = {"·", "█", "▓", "▒", "░", " "};
        for (int y = 0; y < size; y++) {
            StringBuilder line = new StringBuilder();
            for (int x = 0; x < size; x++) {
                int idx = pixels[y][x];
                line.append(idx < chars.length ? chars[idx] : " ");
            }
            System.out.println(line);
        }
    }

    public void run() throws Exception {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        Type dataType = new TypeToken<EditorData>(){}.getType();

        if (loadFile != null) {
            String json = new String(Files.readAllBytes(Paths.get(loadFile)));
            EditorData data = gson.fromJson(json, dataType);
            fromData(data);
        } else if (newFlag || size > 0) {
            initPixels();
        } else {
            size = 32;
            initPixels();
        }

        if (hotspot != null) {
            String[] parts = hotspot.split(",");
            if (parts.length == 2) {
                hotspotPos[0] = Integer.parseInt(parts[0]);
                hotspotPos[1] = Integer.parseInt(parts[1]);
            }
        }

        if (color != null) {
            Map<String, Integer> colorMap = new HashMap<>();
            colorMap.put("#000000", 1); colorMap.put("#FFFFFF", 2);
            colorMap.put("#FF0000", 3); colorMap.put("#00FF00", 4);
            colorMap.put("#0000FF", 5); colorMap.put("#FFFF00", 6);
            colorMap.put("#FF00FF", 7); colorMap.put("#00FFFF", 8);
            currentColor = colorMap.getOrDefault(color.toUpperCase(), 1);
        }

        if (pixel != null) {
            String[] parts = pixel.split(",");
            if (parts.length == 2) {
                int x = Integer.parseInt(parts[0]);
                int y = Integer.parseInt(parts[1]);
                setPixel(x, y, currentColor);
            }
        }

        if (fill != null) {
            String[] parts = fill.split(",");
            if (parts.length == 2) {
                int x = Integer.parseInt(parts[0]);
                int y = Integer.parseInt(parts[1]);
                fill(x, y, currentColor);
            }
        }

        if (saveFile != null) {
            String json = gson.toJson(toData());
            Files.write(Paths.get(saveFile), json.getBytes());
            System.out.println("\u001B[32mПроект сохранён в " + saveFile + "\u001B[0m");
        }

        if (exportFile != null) {
            exportCur(exportFile);
        }

        if (preview) {
            preview();
        }

        if (!newFlag && loadFile == null && saveFile == null && exportFile == null && !preview && pixel == null && fill == null) {
            System.out.println("Используйте --help для справки.");
        }
    }

    public static void main(String[] args) throws Exception {
        CursorEditor editor = new CursorEditor();
        JCommander.newBuilder().addObject(editor).build().parse(args);
        editor.run();
    }
}

File: CursorEditor.kt
// CursorEditor.kt
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.awt.Color
import java.awt.image.BufferedImage
import java.io.File
import javax.imageio.ImageIO

class CursorEditor {
    @Parameter(names = ["--new"])
    private var newFlag: Boolean = false

    @Parameter(names = ["--size"])
    private var size: Int = 32

    @Parameter(names = ["--load"])
    private var loadFile: String? = null

    @Parameter(names = ["--save"])
    private var saveFile: String? = null

    @Parameter(names = ["--export"])
    private var exportFile: String? = null

    @Parameter(names = ["--hotspot"])
    private var hotspot: String? = null

    @Parameter(names = ["--color"])
    private var color: String? = null

    @Parameter(names = ["--pixel"])
    private var pixel: String? = null

    @Parameter(names = ["--fill"])
    private var fill: String? = null

    @Parameter(names = ["--preview"])
    private var preview: Boolean = false

    data class EditorData(val size: Int, val pixels: Array<IntArray>, val hotspot: IntArray, val palette: IntArray)

    private lateinit var pixels: Array<IntArray>
    private var currentColor = 1
    private var hotspotPos = intArrayOf(0, 0)
    private val palette = intArrayOf(1, 2, 3, 4, 5, 6, 7, 8)
    private val colorMap = mapOf(
        0 to intArrayOf(255, 255, 255, 0),
        1 to intArrayOf(0, 0, 0, 255),
        2 to intArrayOf(255, 255, 255, 255),
        3 to intArrayOf(255, 0, 0, 255),
        4 to intArrayOf(0, 255, 0, 255),
        5 to intArrayOf(0, 0, 255, 255),
        6 to intArrayOf(255, 255, 0, 255),
        7 to intArrayOf(255, 0, 255, 255),
        8 to intArrayOf(0, 255, 255, 255)
    )

    private fun initPixels() {
        pixels = Array(size) { IntArray(size) }
    }

    private fun setPixel(x: Int, y: Int, colorIdx: Int) {
        if (x in 0 until size && y in 0 until size) {
            pixels[y][x] = colorIdx
        }
    }

    private fun getPixel(x: Int, y: Int): Int {
        return if (x in 0 until size && y in 0 until size) pixels[y][x] else 0
    }

    private fun fill(x: Int, y: Int, colorIdx: Int) {
        val target = getPixel(x, y)
        if (target == colorIdx) return
        val stack = mutableListOf(Pair(x, y))
        val visited = mutableSetOf<Pair<Int, Int>>()
        while (stack.isNotEmpty()) {
            val (cx, cy) = stack.removeAt(stack.size - 1)
            if (visited.contains(Pair(cx, cy))) continue
            visited.add(Pair(cx, cy))
            if (getPixel(cx, cy) == target) {
                setPixel(cx, cy, colorIdx)
                for ((dx, dy) in listOf(1 to 0, -1 to 0, 0 to 1, 0 to -1)) {
                    val nx = cx + dx
                    val ny = cy + dy
                    if (nx in 0 until size && ny in 0 until size) {
                        stack.add(Pair(nx, ny))
                    }
                }
            }
        }
    }

    private fun toImage(): BufferedImage {
        val img = BufferedImage(size, size, BufferedImage.TYPE_INT_ARGB)
        for (y in 0 until size) {
            for (x in 0 until size) {
                val c = colorMap[pixels[y][x]] ?: intArrayOf(0, 0, 0, 0)
                img.setRGB(x, y, Color(c[0], c[1], c[2], c[3]).rgb)
            }
        }
        return img
    }

    private fun exportCur(filename: String) {
        val img = toImage()
        val pngName = filename.replace(".cur", ".png")
        ImageIO.write(img, "png", File(pngName))
        println("\u001B[32mЭкспортировано в $filename (как PNG)\u001B[0m")
    }

    private fun toData(): EditorData {
        return EditorData(size, pixels, hotspotPos, palette)
    }

    private fun fromData(data: EditorData) {
        this.size = data.size
        this.pixels = data.pixels
        this.hotspotPos = data.hotspot
    }

    private fun preview() {
        println("\u001B[36mПредпросмотр курсора (${size}x${size}), hotspot: (${hotspotPos[0]},${hotspotPos[1]})\u001B[0m")
        val chars = arrayOf("·", "█", "▓", "▒", "░", " ")
        for (y in 0 until size) {
            val line = StringBuilder()
            for (x in 0 until size) {
                val idx = pixels[y][x]
                line.append(if (idx < chars.size) chars[idx] else " ")
            }
            println(line)
        }
    }

    fun run() {
        val gson = GsonBuilder().setPrettyPrinting().create()
        val type = object : TypeToken<EditorData>() {}.type

        if (loadFile != null) {
            val json = File(loadFile).readText()
            val data = gson.fromJson<EditorData>(json, type)
            fromData(data)
        } else if (newFlag || size > 0) {
            initPixels()
        } else {
            size = 32
            initPixels()
        }

        hotspot?.let {
            val parts = it.split(",")
            if (parts.size == 2) {
                hotspotPos[0] = parts[0].toInt()
                hotspotPos[1] = parts[1].toInt()
            }
        }

        color?.let {
            val colorMap = mapOf(
                "#000000" to 1, "#FFFFFF" to 2, "#FF0000" to 3,
                "#00FF00" to 4, "#0000FF" to 5, "#FFFF00" to 6,
                "#FF00FF" to 7, "#00FFFF" to 8
            )
            currentColor = colorMap[it.uppercase()] ?: 1
        }

        pixel?.let {
            val parts = it.split(",")
            if (parts.size == 2) {
                setPixel(parts[0].toInt(), parts[1].toInt(), currentColor)
            }
        }

        fill?.let {
            val parts = it.split(",")
            if (parts.size == 2) {
                fill(parts[0].toInt(), parts[1].toInt(), currentColor)
            }
        }

        saveFile?.let {
            val json = gson.toJson(toData())
            File(it).writeText(json)
            println("\u001B[32mПроект сохранён в $it\u001B[0m")
        }

        exportFile?.let {
            exportCur(it)
        }

        if (preview) {
            preview()
        }

        if (!newFlag && loadFile == null && saveFile == null && exportFile == null && !preview && pixel == null && fill == null) {
            println("Используйте --help для справки.")
        }
    }
}

fun main(args: Array<String>) {
    val editor = CursorEditor()
    JCommander.newBuilder().addObject(editor).build().parse(*args)
    editor.run()
}
